supervisor/supervisor.mk

```diff
@@ -94,9 +94,40 @@ ifndef USB_HID_DEVICES
 USB_HID_DEVICES = "KEYBOARD,MOUSE,CONSUMER,GAMEPAD"
 endif
 
-# SAMD21 needs separate endpoint pairs for MSC BULK IN and BULK OUT, otherwise it's erratic.
-ifndef USB_MSC_NUM_ENDPOINT_PAIRS
-USB_MSC_NUM_ENDPOINT_PAIRS = 1
+ifndef USB_RELATIVE_EP_NUM
+USB_RELATIVE_EP_NUM = 1
+endif
+
+ifndef USB_CDC_EP_NUM_NOTIFICATION
+USB_CDC_EP_NUM_NOTIFICATION = 0
+endif
+
+ifndef USB_CDC_EP_NUM_DATA_OUT
+USB_CDC_EP_NUM_DATA_OUT = 0
+endif
+
+ifndef USB_CDC_EP_NUM_DATA_IN
+USB_CDC_EP_NUM_DATA_IN = 0
+endif
+
+ifndef USB_MSC_EP_NUM_OUT
+USB_MSC_EP_NUM_OUT = 0
+endif
+
+ifndef USB_MSC_EP_NUM_IN
+USB_MSC_EP_NUM_IN = 0
+endif
+
+ifndef USB_HID_EP_NUM_IN
+USB_HID_EP_NUM_IN = 0
+endif
+
+ifndef USB_MIDI_EP_NUM_OUT
+USB_MIDI_EP_NUM_OUT = 0
+endif
+
+ifndef USB_MIDI_EP_NUM_IN
+USB_MIDI_EP_NUM_IN = 0
 endif
```

Comment:
I'm thinking that if any of these values are set, then `USB_RELATIVE_EP_NUM` should be 0, since `join_interfaces()` is either going to renumber interfaces or it isn't. So I don't think you need a separate `--relative_ep_num` argument. 

I think ideally the validation in the scripts would do the following: if any `*_EP_NUM_*` is set, then it should be the case that all used `EP_NUM` values should be set, and it should be an error if they're not. Since not all devices might be enabled, you don't need all to be checked, just the ones that are used. So the check is:

If any `EP_NUM` value is set, then assume absolute `EP_NUM`s. When an `EP_NUM` is used, make sure it's not zero. If a zero `EP_NUM` is used, then that's an error. So just check for that in the script.

This assumes that they are either all absolute or all relative. I don't know if there are any chips that allow some relative and need some fixed, but we can cross that bridge later. It sounds like the chip in question requires all fixed.

Reply:
I thought about that too but it is actually useful to provide the relative ep number before join_interfaces renumbers every because it preserves the relationship. That's why the special case for SAMD21 can be removed.

I agree that without renumbering we should validate that all endpoints used were provided.

Reply:
I have added validation which checks if the addresses of used endpoints are different from 0.